{"data": {"repository": {"pullRequest": {"id": "MDExOlB1bGxSZXF1ZXN0Mzg3MDk1MTQ3", "number": 9022, "reviewThreads": {"totalCount": 24, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xMlQyMToxODoyMlrODnoVHg==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0xMlQwMDoxNjo0N1rOEE5Dhg==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjQyODgxODIyOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xMlQyMToxODoyMlrOF1vnMA==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xMlQyMToxODoyMlrOF1vnMA==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDM5MTg5ODkyOA==", "bodyText": "Can we use package private for now instead?", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r391898928", "createdAt": "2020-03-12T21:18:22Z", "author": {"login": "jianghaolu"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -23,7 +23,7 @@\n  */\n @Immutable\n public class ChainedTokenCredential implements TokenCredential {\n-    private final Deque<TokenCredential> credentials;\n+    protected final Deque<TokenCredential> credentials;", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "56786ea96df01bb03e61c67d31071ebe28fd35f4"}, "originalPosition": 5}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjQyODgyMTQ1OnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xMlQyMToxOToxNlrOF1vpAQ==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xMlQyMToxOToxNlrOF1vpAQ==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDM5MTg5OTM5Mw==", "bodyText": "We have changed this to run in serial: https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java#L46", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r391899393", "createdAt": "2020-03-12T21:19:16Z", "author": {"login": "jianghaolu"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredential.java", "diffHunk": "@@ -39,4 +45,28 @@\n             new SharedTokenCacheCredential(null, \"04b07795-8ddb-461a-bbee-02f9e1bf7b46\",\n                 identityClientOptions))));\n     }\n+\n+    @Override\n+    public Mono<AccessToken> getToken(TokenRequestContext request) {\n+        final AtomicReference<Throwable> cause = new AtomicReference<>();\n+        StringBuilder errorMsg = new StringBuilder();\n+        return Flux.fromIterable(credentials).flatMap(p -> p.getToken(request).onErrorResume(t -> {", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "56786ea96df01bb03e61c67d31071ebe28fd35f4"}, "originalPosition": 26}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjQ0NTUxMjA5OnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/SharedTokenCacheCredential.java", "isResolved": true, "comments": {"totalCount": 2, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xOFQxODowMzoyMFrOF4Q58Q==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xOFQxODoyMzo1M1rOF4Roow==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDM5NDU0MTU1Mw==", "bodyText": "This is removing the instructions for the user as to how they'll resolve/fix the issue. We should keep those instructions so users are updated about how to resolve the issue.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r394541553", "createdAt": "2020-03-18T18:03:20Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/SharedTokenCacheCredential.java", "diffHunk": "@@ -86,22 +85,18 @@\n \n                 if (accounts.size() == 0) {\n                     if (username == null) {\n-                        return Mono.error(new RuntimeException(\"No accounts were discovered in the shared token cache.\"\n-                            + \" To fix, authenticate through tooling supporting azure developer sign on.\"));\n+                        return Mono.error(new RuntimeException(\"SharedTokenCacheCredential authentication unavailable. No accounts were discovered in the cache.\"));", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "753e885d8d49aa14143e475c45c6e9bdee8ca666"}, "originalPosition": 14}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDM5NDU1MzUwNw==", "bodyText": "I removed these more specific error messages from the shared token credential. They aren't always as helpful, especially in this case where \"To fix, authenticate through tooling supporting azure developer sign on.\" is very vague. The end goal as we add more specific credential types for authenticating with specific tools that they will provide error messaging which can be much more specific.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r394553507", "createdAt": "2020-03-18T18:23:53Z", "author": {"login": "schaabs"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/SharedTokenCacheCredential.java", "diffHunk": "@@ -86,22 +85,18 @@\n \n                 if (accounts.size() == 0) {\n                     if (username == null) {\n-                        return Mono.error(new RuntimeException(\"No accounts were discovered in the shared token cache.\"\n-                            + \" To fix, authenticate through tooling supporting azure developer sign on.\"));\n+                        return Mono.error(new RuntimeException(\"SharedTokenCacheCredential authentication unavailable. No accounts were discovered in the cache.\"));", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDM5NDU0MTU1Mw=="}, "originalCommit": {"oid": "753e885d8d49aa14143e475c45c6e9bdee8ca666"}, "originalPosition": 14}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjQ0NTUzMzQ2OnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/SharedTokenCacheCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xOFQxODowOTozMlrOF4RICg==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xOFQxODowOTozMlrOF4RICg==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDM5NDU0NTE2Mg==", "bodyText": "Same here, we should keep the instructions for the user as to how they'll resolve/fix the issue.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r394545162", "createdAt": "2020-03-18T18:09:32Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/SharedTokenCacheCredential.java", "diffHunk": "@@ -86,22 +85,18 @@\n \n                 if (accounts.size() == 0) {\n                     if (username == null) {\n-                        return Mono.error(new RuntimeException(\"No accounts were discovered in the shared token cache.\"\n-                            + \" To fix, authenticate through tooling supporting azure developer sign on.\"));\n+                        return Mono.error(new RuntimeException(\"SharedTokenCacheCredential authentication unavailable. No accounts were discovered in the cache.\"));\n                     } else {\n-                        return Mono.error(new RuntimeException(String.format(\"User account '%s' was not found in the \"\n-                            + \"shared token cache. Discovered Accounts: [ '%s' ]\", username, accounts.values().stream()\n-                            .map(IAccount::username).collect(Collectors.joining(\", \")))));\n+                        return Mono.error(new RuntimeException(String.format(\"SharedTokenCacheCredential authentication unavailable. No account \"\n+                            + \"matching the specified %s was found in the cache.\", username)));\n                     }\n                 } else if (accounts.size() > 1) {\n                     if (username == null) {\n-                        return Mono.error(new RuntimeException(\"Multiple accounts were discovered in the shared token \"\n-                            + \"cache. To fix, set the AZURE_USERNAME environment variable to the preferred username, \"\n-                            + \"or specify it when constructing SharedTokenCacheCredential.\"));\n+                        return Mono.error(new RuntimeException(\"SharedTokenCacheCredential authentication unavailable. \"\n+                            + \"Multiple accounts were found in the cache. Use username and tenant id to disambiguate.\"));", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "753e885d8d49aa14143e475c45c6e9bdee8ca666"}, "originalPosition": 28}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjQ0NTUzNTU3OnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/SharedTokenCacheCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xOFQxODoxMDoxNVrOF4RJfg==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xOFQxODoxMDoxNVrOF4RJfg==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDM5NDU0NTUzNA==", "bodyText": "Same here, we should keep the update the user why the error is thrown and the feature is not supported.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r394545534", "createdAt": "2020-03-18T18:10:15Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/SharedTokenCacheCredential.java", "diffHunk": "@@ -86,22 +85,18 @@\n \n                 if (accounts.size() == 0) {\n                     if (username == null) {\n-                        return Mono.error(new RuntimeException(\"No accounts were discovered in the shared token cache.\"\n-                            + \" To fix, authenticate through tooling supporting azure developer sign on.\"));\n+                        return Mono.error(new RuntimeException(\"SharedTokenCacheCredential authentication unavailable. No accounts were discovered in the cache.\"));\n                     } else {\n-                        return Mono.error(new RuntimeException(String.format(\"User account '%s' was not found in the \"\n-                            + \"shared token cache. Discovered Accounts: [ '%s' ]\", username, accounts.values().stream()\n-                            .map(IAccount::username).collect(Collectors.joining(\", \")))));\n+                        return Mono.error(new RuntimeException(String.format(\"SharedTokenCacheCredential authentication unavailable. No account \"\n+                            + \"matching the specified %s was found in the cache.\", username)));\n                     }\n                 } else if (accounts.size() > 1) {\n                     if (username == null) {\n-                        return Mono.error(new RuntimeException(\"Multiple accounts were discovered in the shared token \"\n-                            + \"cache. To fix, set the AZURE_USERNAME environment variable to the preferred username, \"\n-                            + \"or specify it when constructing SharedTokenCacheCredential.\"));\n+                        return Mono.error(new RuntimeException(\"SharedTokenCacheCredential authentication unavailable. \"\n+                            + \"Multiple accounts were found in the cache. Use username and tenant id to disambiguate.\"));\n                     } else {\n-                        return Mono.error(new RuntimeException(\"Multiple entries for the user account \" + username\n-                            + \" were found in the shared token cache. This is not currently supported by the\"\n-                            + \" SharedTokenCacheCredential.\"));\n+                        return Mono.error(new RuntimeException(String.format(\"SharedTokenCacheCredential authentication unavailable. Multiple accounts matching the specified \"\n+                            + \" %s were found in the cache.\", username)));", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "753e885d8d49aa14143e475c45c6e9bdee8ca666"}, "originalPosition": 34}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjQ0NTU0MTI5OnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xOFQxODoxMTo1NlrOF4RNfg==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xOFQxODoxMTo1NlrOF4RNfg==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDM5NDU0NjU1OA==", "bodyText": "code indentation needs to be fixed.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r394546558", "createdAt": "2020-03-18T18:11:56Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -35,19 +35,15 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        AtomicReference<Throwable> cause = new AtomicReference<>();\n-        return Flux.fromIterable(credentials)\n-            .flatMap(p -> p.getToken(request).onErrorResume(t -> {\n-                if (cause.get() != null) {\n-                    t.initCause(cause.get());\n-                }\n-                cause.set(t);\n-                return Mono.empty();\n-            }))\n+        final StringBuilder errorMsg = new StringBuilder();\n+        return Flux.fromIterable(credentials).flatMap(p -> p.getToken(request).onErrorResume(t -> {", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "753e885d8d49aa14143e475c45c6e9bdee8ca666"}, "originalPosition": 32}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjQ0NTU0NjMyOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xOFQxODoxMzoxOFrOF4RQwg==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wMy0xOFQxODoxMzoxOFrOF4RQwg==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDM5NDU0NzM5NA==", "bodyText": "This error message is not properly formatted, spacing issues will show up before 'authentication failed' text.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r394547394", "createdAt": "2020-03-18T18:13:18Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -35,19 +35,15 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        AtomicReference<Throwable> cause = new AtomicReference<>();\n-        return Flux.fromIterable(credentials)\n-            .flatMap(p -> p.getToken(request).onErrorResume(t -> {\n-                if (cause.get() != null) {\n-                    t.initCause(cause.get());\n-                }\n-                cause.set(t);\n-                return Mono.empty();\n-            }))\n+        final StringBuilder errorMsg = new StringBuilder();\n+        return Flux.fromIterable(credentials).flatMap(p -> p.getToken(request).onErrorResume(t -> {\n+            if (t.getMessage() != null && !t.getMessage().contains(\"authentication unavailable\")) {\n+                throw new RuntimeException(UnavailableError+p.getClass().getSimpleName()+\"authentication failed.\",t);", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "753e885d8d49aa14143e475c45c6e9bdee8ca666"}, "originalPosition": 34}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjU5NDU4NTEyOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialUnavailableException.java", "isResolved": true, "comments": {"totalCount": 3, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNC0yOVQwMTowMzo1NVrOGNsgAw==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNS0xMVQxODowMDo1MlrOGTnTvw==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxMzc2Mw==", "bodyText": "This should extend ClientAuthenticationException.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r417013763", "createdAt": "2020-04-29T01:03:55Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialUnavailableException.java", "diffHunk": "@@ -5,21 +5,20 @@\n \n import com.azure.core.credential.AccessToken;\n import com.azure.core.credential.TokenCredential;\n-import com.azure.core.exception.ClientAuthenticationException;\n \n /**\n  * The exception thrown when a {@link TokenCredential} did not attempt to authenticate and retrieve {@link AccessToken},\n  * as its prerequisite information or state was not available.\n  */\n-public class CredentialUnavailableException extends ClientAuthenticationException {\n+public class CredentialUnavailableException extends RuntimeException {", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 11}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAyMzg0Mw==", "bodyText": "If extend ClientAuthenticationException, There is a bug when running defaultazurecredential.\n\nrun code as follow  when all credentials are unavaliable in defaultazurecredential\npublic static void main(String[] args) {\ntry{\nDefaultAzureCredential defaultAzureCredential=new DefaultAzureCredentialBuilder().build();\nBlobServiceClient blobServiceClient = new BlobServiceClientBuilder()\n.endpoint(endpoint)\n.credential(defaultAzureCredential)\n.buildClient();\nblobServiceClient.createBlobContainer(\"mycontainer\");\n}catch(Exception e){\ne.printStackTrace();\n}\n}", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r417023843", "createdAt": "2020-04-29T01:46:58Z", "author": {"login": "Luyunmt"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialUnavailableException.java", "diffHunk": "@@ -5,21 +5,20 @@\n \n import com.azure.core.credential.AccessToken;\n import com.azure.core.credential.TokenCredential;\n-import com.azure.core.exception.ClientAuthenticationException;\n \n /**\n  * The exception thrown when a {@link TokenCredential} did not attempt to authenticate and retrieve {@link AccessToken},\n  * as its prerequisite information or state was not available.\n  */\n-public class CredentialUnavailableException extends ClientAuthenticationException {\n+public class CredentialUnavailableException extends RuntimeException {", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxMzc2Mw=="}, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 11}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQyMzIyMDE1OQ==", "bodyText": "I am not able to reproduce this when used DefaultAzureCredential when no credential available with KeyVault Secrets client.\nAre you running the above code of Storage with tracing enabled ?", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r423220159", "createdAt": "2020-05-11T18:00:52Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialUnavailableException.java", "diffHunk": "@@ -5,21 +5,20 @@\n \n import com.azure.core.credential.AccessToken;\n import com.azure.core.credential.TokenCredential;\n-import com.azure.core.exception.ClientAuthenticationException;\n \n /**\n  * The exception thrown when a {@link TokenCredential} did not attempt to authenticate and retrieve {@link AccessToken},\n  * as its prerequisite information or state was not available.\n  */\n-public class CredentialUnavailableException extends ClientAuthenticationException {\n+public class CredentialUnavailableException extends RuntimeException {", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxMzc2Mw=="}, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 11}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjU5NDU4NTMzOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 2, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNC0yOVQwMTowMzo1OVrOGNsgHQ==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wMlQxNjozNjo0NFrOGd6WJQ==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxMzc4OQ==", "bodyText": "What will 'throw' achieve here ?", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r417013789", "createdAt": "2020-04-29T01:03:59Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n-                   .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                   .flatMap(p -> p.getToken(request).onErrorResume(Exception.class, t -> {\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           throw logger.logExceptionAsError(new CredentialUnavailableException(", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 38}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNDAxNzgyOQ==", "bodyText": "In the reactive spec, this should return normally rather than throwing (ie. return Mono.error(logger.logExceptionAsError)).\nhttps://github.com/reactive-streams/reactive-streams-jvm#2.13", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r434017829", "createdAt": "2020-06-02T16:36:44Z", "author": {"login": "conniey"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n-                   .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                   .flatMap(p -> p.getToken(request).onErrorResume(Exception.class, t -> {\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           throw logger.logExceptionAsError(new CredentialUnavailableException(", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxMzc4OQ=="}, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 38}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjU5NDU4NTQwOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNC0yOVQwMTowNDowMlrOGNsgKg==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNC0yOVQwMTowNDowMlrOGNsgKg==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxMzgwMg==", "bodyText": "This logic is not correct, we only want to catch CredentialUnavailableException here.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r417013802", "createdAt": "2020-04-29T01:04:02Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n-                   .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                   .flatMap(p -> p.getToken(request).onErrorResume(Exception.class, t -> {", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 36}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjU5NDU4NTU5OnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNC0yOVQwMTowNDoxMVrOGNsgRQ==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNC0yOVQwMTowNDoxMVrOGNsgRQ==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxMzgyOQ==", "bodyText": "This should throw ClientAuth Exception.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r417013829", "createdAt": "2020-04-29T01:04:11Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java", "diffHunk": "@@ -219,10 +218,10 @@\n                         throw logger.logExceptionAsError(\n                             new CredentialUnavailableException(\"Please run 'az login' to set up account\"));\n                     }\n-                    throw logger.logExceptionAsError(new ClientAuthenticationException(redactedOutput, null));\n+                    throw logger.logExceptionAsError(new CredentialUnavailableException(redactedOutput));", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 13}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjU5NDU4NTYxOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNC0yOVQwMTowNDoxMlrOGNsgSQ==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNC0yOVQwMTowNDoxMlrOGNsgSQ==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxMzgzMw==", "bodyText": "This should throw Client Auth Exception.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r417013833", "createdAt": "2020-04-29T01:04:12Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java", "diffHunk": "@@ -219,10 +218,10 @@\n                         throw logger.logExceptionAsError(\n                             new CredentialUnavailableException(\"Please run 'az login' to set up account\"));\n                     }\n-                    throw logger.logExceptionAsError(new ClientAuthenticationException(redactedOutput, null));\n+                    throw logger.logExceptionAsError(new CredentialUnavailableException(redactedOutput));\n                 } else {\n                     throw logger.logExceptionAsError(\n-                        new ClientAuthenticationException(\"Failed to invoke Azure CLI \", null));\n+                        new CredentialUnavailableException(\"Failed to invoke Azure CLI \"));", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 17}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjU5NDYxMzkyOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNC0yOVQwMToyMToyMVrOGNswBg==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNC0yOVQwMToyMToyMVrOGNswBg==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxNzg2Mg==", "bodyText": "We seem to be double logging the exception here.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r417017862", "createdAt": "2020-04-29T01:21:21Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n-                   .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                   .flatMap(p -> p.getToken(request).onErrorResume(Exception.class, t -> {\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           throw logger.logExceptionAsError(new CredentialUnavailableException(", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 38}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjU5NDYxNjgxOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 3, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNC0yOVQwMToyMzowMlrOGNsxjg==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wM1QwNzozOTowM1rOGePw0Q==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxODI1NA==", "bodyText": "A single large message will be unfriendly to read.\nAre we doing this across all languages ?\nBetter to keep the individual exceptions and their stack trace separate", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r417018254", "createdAt": "2020-04-29T01:23:02Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n-                   .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                   .flatMap(p -> p.getToken(request).onErrorResume(Exception.class, t -> {\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           throw logger.logExceptionAsError(new CredentialUnavailableException(\n+                                   unavailableError + p.getClass().getSimpleName() + \" authentication failed.\",\n+                                   t));\n+                       }\n+                       message.append(t.getMessage()).append(\" \"); ", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 42}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQyNDEzMjI0MQ==", "bodyText": "In this pr, I change java error message to java(new), and other languages' error message is now as follow, Do java need change it to java(new)? If you think java need change and the java(new) error message is not good enough, please  change it in the under table , and I will implement.\n\n\n\nScenario\nPython/C#\nJava(Current)\nJava(new)\n\n\n\n\nCredential   Unavailable\nDefaultAzureCredential   failed to retrieve a token from the included credentials.     EnvironmentCredential authentication unavailable. Environment variables are   not fully configured.     ManagedIdentityCredential authentication unavailable. No Managed Identity   endpoint found.     SharedTokenCacheCredential authentication unavailable. No accounts were   found in the cache.\nTried   EnvironmentCredential, ManagedIdentityCredential, SharedTokenCacheCredential,   AzureCliCredential but failed to acquire a token for any of them. Please   verify the environment for the credentials and see more details in the causes   below.\nDefaultAzureCredential failed to retrieve a token from the   included credentials.EnvironmentCredential   authentication unavailable. Environment variables are not fully configured.   ManagedIdentityCredential authentication unavailable. Connection to IMDS   endpoint cannot be established, connect timed out. SharedTokenCacheCredential   authentication unavailable. No accounts were found in the cache.   AzureCliCredential authentication unavailable.\n\n\nAuthenticate   Failed\nDefaultAzureCredential   authentication failed. ---> Azure.Identity.AuthenticationFailedException:   ClientSecretCredential authentication failed. --->   Azure.RequestFailedException: Service request failed.     Status: 400 (Bad Request)   Content:       {\"error\":\"unauthorized_client\",\"error_description\":\"AADSTS700016:   Application with identifier '78e8f606-0950-4210-906d-1d321511c09d' was not   found in the directory 'a7fc734e-9961-43ce-b4de-21b8b38403ba'. This can   happen if the application has not been installed by the administrator of the   tenant or consented to by any user in the tenant. You may have sent your   authentication request to the wrong tenant.\\r\\nTrace ID:   9f3624a7-49ca-4097-a339-780fd4e41600\\r\\nCorrelation ID:   b1acfa40-06cc-42a1-a2f1-d35b94db3628\\r\\nTimestamp: 2020-02-06   01:13:33Z\",\"error_codes\":[700016],\"timestamp\":\"2020-02-06   01:13:33Z\",\"trace_id\":\"9f3624a7-49ca-4097-a339-780fd4e41600\",\"correlation_id\":\"b1acfa40-06cc-42a1-a2f1-d35b94db3628\",\"error_uri\":\"https://login.microsoftonline.com/error?code=700016\"}\nTried   EnvironmentCredential, ManagedIdentityCredential, SharedTokenCacheCredential,   AzureCliCredential but failed to acquire a token for any of them. Please   verify the environment for the credentials and see more details in the causes   below.\nDefaultAzureCredential   authentication failed. ---> EnvironmentCredential authentication failed.   \u2026\u2026.   \u00a0Suppressed:   com.microsoft.aad.msal4j.MsalServiceException: AADSTS90002: Tenant   '72f988bf-86f1-41af-91ab-2d7cd011db41' not found. This may happen if there   are no active subscriptions for the tenant. Check to make sure you have the   correct tenant ID. Check with your subscription administrator.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r424132241", "createdAt": "2020-05-13T02:06:01Z", "author": {"login": "Luyunmt"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n-                   .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                   .flatMap(p -> p.getToken(request).onErrorResume(Exception.class, t -> {\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           throw logger.logExceptionAsError(new CredentialUnavailableException(\n+                                   unavailableError + p.getClass().getSimpleName() + \" authentication failed.\",\n+                                   t));\n+                       }\n+                       message.append(t.getMessage()).append(\" \"); ", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxODI1NA=="}, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 42}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNDM2ODcyMQ==", "bodyText": "We need to retain the stack traces from individual exception and chain them.\nJava devs use the stack trace to analyze / debug issues.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r434368721", "createdAt": "2020-06-03T07:39:03Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n-                   .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                   .flatMap(p -> p.getToken(request).onErrorResume(Exception.class, t -> {\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           throw logger.logExceptionAsError(new CredentialUnavailableException(\n+                                   unavailableError + p.getClass().getSimpleName() + \" authentication failed.\",\n+                                   t));\n+                       }\n+                       message.append(t.getMessage()).append(\" \"); ", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQxNzAxODI1NA=="}, "originalCommit": {"oid": "6b25f41ea333a41910ae1af960e71654f2a9f368"}, "originalPosition": 42}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjcwMzMxMjI5OnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/AzureCliCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wMlQxNTo1MToyM1rOGd4Rig==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wMlQxNTo1MToyM1rOGd4Rig==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzMzk4Mzg4Mg==", "bodyText": "I believe onErrorMap would be the more appropriate operator to use here since this appear to be more of a mapping operation than resuming (since it just triggers the error stream again).", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r433983882", "createdAt": "2020-06-02T15:51:23Z", "author": {"login": "alzimmermsft"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/AzureCliCredential.java", "diffHunk": "@@ -32,6 +32,8 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        return identityClient.authenticateWithAzureCli(request);\n+        return identityClient.authenticateWithAzureCli(request)\n+                .onErrorResume(t -> Mono.error(new CredentialUnavailableException(", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "38009fd35dea903e40b687e2e0d3f84fe2eb57f6"}, "originalPosition": 6}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjcwMzMyNjQwOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 2, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wMlQxNTo1Mzo0MFrOGd4arA==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wMlQxNjozOToyN1rOGd6dGQ==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzMzk4NjIyMA==", "bodyText": "This should use Reactor's Mono.error instead. Having a throw statement within a reactive stream will cause consumers to need to both catch using try/catch in the code block and add in a Reactor error operator such as onError.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r433986220", "createdAt": "2020-06-02T15:53:40Z", "author": {"login": "alzimmermsft"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n                    .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           throw logger.logExceptionAsError(new CredentialUnavailableException(", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "38009fd35dea903e40b687e2e0d3f84fe2eb57f6"}, "originalPosition": 37}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNDAxOTYwOQ==", "bodyText": "Reactor actually handles the case where downstream subscribers throw on flatmap. It'll wrap it and propagate a downstream Mono.error. In general though, we should not explicitly throw.\nhttps://github.com/reactive-streams/reactive-streams-jvm#2.13", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r434019609", "createdAt": "2020-06-02T16:39:27Z", "author": {"login": "conniey"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n                    .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           throw logger.logExceptionAsError(new CredentialUnavailableException(", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzMzk4NjIyMA=="}, "originalCommit": {"oid": "38009fd35dea903e40b687e2e0d3f84fe2eb57f6"}, "originalPosition": 37}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjcwMzMzNzk5OnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wMlQxNTo1NTozMlrOGd4iJA==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wMlQxNTo1NTozMlrOGd4iJA==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzMzk4ODEzMg==", "bodyText": "This code confuses me a bit, first we resume the error, potentially throwing or resuming with empty, then if it is the latter of empty we return an error. Could this be streamlined at all in the earlier code to bypass this operator? Such as using onErrorMap above.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r433988132", "createdAt": "2020-06-02T15:55:32Z", "author": {"login": "alzimmermsft"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n                    .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           throw logger.logExceptionAsError(new CredentialUnavailableException(\n+                                   unavailableError + p.getClass().getSimpleName() + \" authentication failed.\",\n+                                   t));\n+                       }\n+                       message.append(t.getMessage()).append(\" \"); \n                        return Mono.empty();\n                    }), 1)\n                    .next()\n                    .switchIfEmpty(Mono.defer(() -> {\n-\n-                       StringBuilder message = new StringBuilder(\"Tried \"\n-                             + credentials.stream().map(c -> c.getClass().getSimpleName())\n-                                   .collect(Collectors.joining(\", \"))\n-                             + \" but failed to acquire a token for any of them. Please verify the\"\n-                             + \" environment for the credentials\"\n-                             + \" and see more details in the causes below.\");\n-\n-                       // Chain Exceptions.\n-                       CredentialUnavailableException last = exceptions.get(exceptions.size() - 1);\n-                       for (int z = exceptions.size() - 2; z >= 0; z--) {\n-                           CredentialUnavailableException current = exceptions.get(z);\n-                           last = new CredentialUnavailableException(current.getMessage(), last);\n-                       }\n-                       return Mono.error(new CredentialUnavailableException(message.toString(), last));\n+                       return Mono.error(new CredentialUnavailableException(message.toString()));", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "38009fd35dea903e40b687e2e0d3f84fe2eb57f6"}, "originalPosition": 61}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjcwMzM0MzYyOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wMlQxNTo1NjoyOVrOGd4l5g==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wMlQxNTo1NjoyOVrOGd4l5g==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzMzk4OTA5NA==", "bodyText": "onErrorMap?", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r433989094", "createdAt": "2020-06-02T15:56:29Z", "author": {"login": "alzimmermsft"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java", "diffHunk": "@@ -822,7 +822,8 @@ private HttpPipeline setupPipeline(HttpClient httpClient) {\n             }\n \n             return true;\n-        });\n+        }).onErrorResume(t -> Mono.error(new CredentialUnavailableException(", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "38009fd35dea903e40b687e2e0d3f84fe2eb57f6"}, "originalPosition": 5}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjcwMzM0OTQxOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/test/java/com/azure/identity/AzureCliCredentialTest.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wMlQxNTo1NzoxOFrOGd4pIQ==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wMlQxNTo1NzoxOFrOGd4pIQ==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzMzk4OTkyMQ==", "bodyText": "Does the actual error message have an additional space? Could we clean that up if so.\n\n  \n    \n  \n    \n\n  \n  This file contains bidirectional Unicode text that may be interpreted or compiled differently than what appears below. To review, open the file in an editor that reveals hidden Unicode characters. Learn more about bidirectional Unicode characters\n\n\n  \n\n\n    \n      \n        Suggested change\n        \n          \n    \n\n        \n      \n    \n    \n      \n          \n            \n                        .expectErrorMatches(e -> e instanceof CredentialUnavailableException && e.getMessage().contains(\"Azure  not Login\"))\n          \n          \n            \n                        .expectErrorMatches(e -> e instanceof CredentialUnavailableException && e.getMessage().contains(\"Azure not Login\"))", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r433989921", "createdAt": "2020-06-02T15:57:18Z", "author": {"login": "alzimmermsft"}, "path": "sdk/identity/azure-identity/src/test/java/com/azure/identity/AzureCliCredentialTest.java", "diffHunk": "@@ -77,7 +77,7 @@ public void azureCliCredentialAzNotLogInException() throws Exception {\n         // test\n         AzureCliCredential credential = new AzureCliCredentialBuilder().build();\n         StepVerifier.create(credential.getToken(request))\n-            .expectErrorMatches(e -> e instanceof Exception && \"Azure  not Login\".equals(e.getMessage()))\n+            .expectErrorMatches(e -> e instanceof CredentialUnavailableException && e.getMessage().contains(\"Azure  not Login\"))", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "38009fd35dea903e40b687e2e0d3f84fe2eb57f6"}, "originalPosition": 14}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjcwNTcxMDU0OnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/AzureCliCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wM1QwNzozNzoxNVrOGePs_A==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wM1QwNzozNzoxNVrOGePs_A==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNDM2Nzc0MA==", "bodyText": "This onErrorMap should be removed.\nWe don't want to map any exception to CredentialUnavailableException.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r434367740", "createdAt": "2020-06-03T07:37:15Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/AzureCliCredential.java", "diffHunk": "@@ -32,6 +32,8 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        return identityClient.authenticateWithAzureCli(request);\n+        return identityClient.authenticateWithAzureCli(request)\n+                .onErrorMap(t -> new CredentialUnavailableException(", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "e282c818792b0e125427916dba8c4ee73b511c3b"}, "originalPosition": 6}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjcwNTcxMzA3OnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 3, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wM1QwNzozNzo1NlrOGePugQ==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wNVQwNDo0OTo0NFrOGfgTEA==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNDM2ODEyOQ==", "bodyText": "This if statement logic is redundant now.\ncan be removed.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r434368129", "createdAt": "2020-06-03T07:37:56Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n                    .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "e282c818792b0e125427916dba8c4ee73b511c3b"}, "originalPosition": 36}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNDQwNDI1Ng==", "bodyText": "In .net python and js, if authenticate unavaliable, continuing to run follow credential. If authenticate  failed, stop running and print the failed credential and its failed reason.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r434404256", "createdAt": "2020-06-03T08:40:49Z", "author": {"login": "Luyunmt"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n                    .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNDM2ODEyOQ=="}, "originalCommit": {"oid": "e282c818792b0e125427916dba8c4ee73b511c3b"}, "originalPosition": 36}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNTY4ODIwOA==", "bodyText": "Yeah, the onErrorResume will only get invoked if CredentialUnavailabeException is thrown.\nA non CredentialUnavailabeException will never reach this block and will automatically break the chain.\nSo, this check in the if statement for a non CredentialUnavailabeException is not needed as a non CredentialUnavailabeException will never enter this block now.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r435688208", "createdAt": "2020-06-05T04:49:44Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n                    .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNDM2ODEyOQ=="}, "originalCommit": {"oid": "e282c818792b0e125427916dba8c4ee73b511c3b"}, "originalPosition": 36}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjcwNTcxODAyOnYy", "diffSide": "LEFT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wM1QwNzozOToyOVrOGePxoQ==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wM1QwNzozOToyOVrOGePxoQ==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNDM2ODkyOQ==", "bodyText": "We need to retain the stack traces from individual exception and chain them.\nJava devs use the stack trace to analyze / debug issues.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r434368929", "createdAt": "2020-06-03T07:39:29Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -36,29 +38,20 @@\n \n     @Override\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n+        StringBuilder message = new StringBuilder();\n         return Flux.fromIterable(credentials)\n                    .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                        return Mono.error(logger.logExceptionAsError(new CredentialUnavailableException(\n+                                   unavailableError + p.getClass().getSimpleName() + \" authentication failed.\",\n+                                   t)));\n+                       }\n+                       message.append(t.getMessage()).append(\" \"); \n                        return Mono.empty();\n                    }), 1)\n                    .next()\n                    .switchIfEmpty(Mono.defer(() -> {\n-\n-                       StringBuilder message = new StringBuilder(\"Tried \"\n-                             + credentials.stream().map(c -> c.getClass().getSimpleName())\n-                                   .collect(Collectors.joining(\", \"))\n-                             + \" but failed to acquire a token for any of them. Please verify the\"\n-                             + \" environment for the credentials\"\n-                             + \" and see more details in the causes below.\");\n-\n-                       // Chain Exceptions.\n-                       CredentialUnavailableException last = exceptions.get(exceptions.size() - 1);\n-                       for (int z = exceptions.size() - 2; z >= 0; z--) {", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "e282c818792b0e125427916dba8c4ee73b511c3b"}, "originalPosition": 56}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjcxMzg4OTEwOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 3, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wNVQwNDo1MToyN1rOGfgUoQ==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0wOFQwMjo1Mzo1OFrOGgN62g==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNTY4ODYwOQ==", "bodyText": "WHat does the error message with stack trace look like now ?\nCan you post the sample output here ?", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r435688609", "createdAt": "2020-06-05T04:51:27Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -39,26 +40,24 @@\n         List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n         return Flux.fromIterable(credentials)\n                    .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           return Mono.error(new CredentialUnavailableException(\n+                                   unavailableError + p.getClass().getSimpleName() + \" authentication failed.\",\n+                                   t));\n+                       }\n                        exceptions.add(t);\n                        return Mono.empty();\n                    }), 1)\n                    .next()\n                    .switchIfEmpty(Mono.defer(() -> {\n-\n-                       StringBuilder message = new StringBuilder(\"Tried \"\n-                             + credentials.stream().map(c -> c.getClass().getSimpleName())\n-                                   .collect(Collectors.joining(\", \"))\n-                             + \" but failed to acquire a token for any of them. Please verify the\"\n-                             + \" environment for the credentials\"\n-                             + \" and see more details in the causes below.\");\n-\n                        // Chain Exceptions.\n                        CredentialUnavailableException last = exceptions.get(exceptions.size() - 1);\n                        for (int z = exceptions.size() - 2; z >= 0; z--) {\n                            CredentialUnavailableException current = exceptions.get(z);\n-                           last = new CredentialUnavailableException(current.getMessage(), last);\n+                           last = new CredentialUnavailableException(current.getMessage() + \" \" + last.getMessage(),\n+                                last.getCause());\n                        }\n-                       return Mono.error(new CredentialUnavailableException(message.toString(), last));\n+                       return Mono.error(last);", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "ac8295bf774dec5959a0692f356f60eb36336a1e"}, "originalPosition": 48}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNjQzNTU4MQ==", "bodyText": "Exception in thread \"main\" java.lang.RuntimeException: Max retries 3 times exceeded. Error Details: EnvironmentCredential authentication unavailable. Environment variables are not fully configured.\nManagedIdentityCredential authentication unavailable. Connection to IMDS endpoint cannot be established, Network is unreachable: connect.\nSharedTokenCacheCredential authentication unavailable. No accounts were found in the cache.\nIntelliJ Authentication not available. Please log in with Azure Tools for IntelliJ plugin in the IDE.\nFailed to read Vs Code credentials from Windows Credential API.\nAzureCliCredential authentication unavailable. Azure CLI not installed\nat com.azure.core.http.policy.RetryPolicy.lambda$attemptAsync$1(RetryPolicy.java:119)\nat reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:88)\nat reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)\nat reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)\nat reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:185)\nat reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:251)\nat reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onError(MonoPeekTerminal.java:251)\nat reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onError(MonoPeekTerminal.java:251)\nat reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onError(FluxPeekFuseable.java:894)\nat reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onError(FluxPeekFuseable.java:894)\nat reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onError(Operators.java:1994)\nat reactor.core.publisher.Operators.error(Operators.java:196)\nat reactor.core.publisher.MonoError.subscribe(MonoError.java:52)\nat reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)\nat reactor.core.publisher.Mono.subscribe(Mono.java:4218)\nat reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onComplete(FluxSwitchIfEmpty.java:75)\nat reactor.core.publisher.MonoNext$NextSubscriber.onComplete(MonoNext.java:96)\nat reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:838)\nat reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:600)\nat reactor.core.publisher.FluxFlatMap$FlatMapMain.innerComplete(FluxFlatMap.java:909)\nat reactor.core.publisher.FluxFlatMap$FlatMapInner.onComplete(FluxFlatMap.java:1013)\nat reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onComplete(Operators.java:1989)\nat reactor.core.publisher.Operators.complete(Operators.java:135)\nat reactor.core.publisher.MonoEmpty.subscribe(MonoEmpty.java:45)\nat reactor.core.publisher.Mono.subscribe(Mono.java:4218)\nat reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:97)\nat reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:126)\nat reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onError(Operators.java:1994)\nat reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:135)\nat reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:73)\nat reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1755)\nat reactor.core.publisher.MonoCompletionStage.lambda$subscribe$0(MonoCompletionStage.java:86)\nat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)\nat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)\nat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)\nat java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705)\nat java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1692)\nat java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)\nat java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)\nat java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)\nat java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)\nat java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)\nSuppressed: java.lang.Exception: #block terminated with an error\nat reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)\nat reactor.core.publisher.Mono.block(Mono.java:1678)\nat com.azure.security.keyvault.secrets.SecretClient.setSecretWithResponse(SecretClient.java:113)\nat com.azure.security.keyvault.secrets.SecretClient.setSecret(SecretClient.java:93)\nat com.azure.identity.Test.main(Test.java:31)\nCaused by: com.azure.identity.CredentialUnavailableException: EnvironmentCredential authentication unavailable. Environment variables are not fully configured.\nManagedIdentityCredential authentication unavailable. Connection to IMDS endpoint cannot be established, Network is unreachable: connect.\nSharedTokenCacheCredential authentication unavailable. No accounts were found in the cache.\nIntelliJ Authentication not available. Please log in with Azure Tools for IntelliJ plugin in the IDE.\nFailed to read Vs Code credentials from Windows Credential API.\nAzureCliCredential authentication unavailable. Azure CLI not installed\nat com.azure.identity.ChainedTokenCredential.lambda$2(ChainedTokenCredential.java:58)\nat reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:44)\n... 28 more", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r436435581", "createdAt": "2020-06-08T02:53:15Z", "author": {"login": "Luyunmt"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -39,26 +40,24 @@\n         List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n         return Flux.fromIterable(credentials)\n                    .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           return Mono.error(new CredentialUnavailableException(\n+                                   unavailableError + p.getClass().getSimpleName() + \" authentication failed.\",\n+                                   t));\n+                       }\n                        exceptions.add(t);\n                        return Mono.empty();\n                    }), 1)\n                    .next()\n                    .switchIfEmpty(Mono.defer(() -> {\n-\n-                       StringBuilder message = new StringBuilder(\"Tried \"\n-                             + credentials.stream().map(c -> c.getClass().getSimpleName())\n-                                   .collect(Collectors.joining(\", \"))\n-                             + \" but failed to acquire a token for any of them. Please verify the\"\n-                             + \" environment for the credentials\"\n-                             + \" and see more details in the causes below.\");\n-\n                        // Chain Exceptions.\n                        CredentialUnavailableException last = exceptions.get(exceptions.size() - 1);\n                        for (int z = exceptions.size() - 2; z >= 0; z--) {\n                            CredentialUnavailableException current = exceptions.get(z);\n-                           last = new CredentialUnavailableException(current.getMessage(), last);\n+                           last = new CredentialUnavailableException(current.getMessage() + \" \" + last.getMessage(),\n+                                last.getCause());\n                        }\n-                       return Mono.error(new CredentialUnavailableException(message.toString(), last));\n+                       return Mono.error(last);", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNTY4ODYwOQ=="}, "originalCommit": {"oid": "ac8295bf774dec5959a0692f356f60eb36336a1e"}, "originalPosition": 48}, {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNjQzNTY3NA==", "bodyText": "if failed\uff1a\nException in thread \"main\" java.lang.RuntimeException: Max retries 3 times exceeded. Error Details: DefaultAzureCredential authentication failed. ---> EnvironmentCredential authentication failed. Error Details:AADSTS90002: Tenant '72f988bf-86f1-41af-91ab-2d7cd011db46' not found. This may happen if there are no active subscriptions for the tenant. Check to\nmake sure you have the correct tenant ID. Check with your subscription administrator.\nTrace ID: 43ffc9ad-f468-4665-ac3a-14a278a98600\nCorrelation ID: e3d3e754-89f0-4839-870c-57e6aba9a5f0\nTimestamp: 2020-06-08 01:49:43Z\nat com.azure.core.http.policy.RetryPolicy.lambda$attemptAsync$1(RetryPolicy.java:119)\nat reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:88)\nat reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)\nat reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)\nat reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:185)\nat reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:251)\nat reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onError(MonoPeekTerminal.java:251)\nat reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onError(MonoPeekTerminal.java:251)\nat reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onError(FluxPeekFuseable.java:894)\nat reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onError(FluxPeekFuseable.java:894)\nat reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onError(Operators.java:1994)\nat reactor.core.publisher.MonoNext$NextSubscriber.onError(MonoNext.java:87)\nat reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:834)\nat reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:600)\nat reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:580)\nat reactor.core.publisher.FluxFlatMap$FlatMapMain.innerError(FluxFlatMap.java:855)\nat reactor.core.publisher.FluxFlatMap$FlatMapInner.onError(FluxFlatMap.java:1006)\nat reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:100)\nat reactor.core.publisher.Operators.error(Operators.java:196)\nat reactor.core.publisher.MonoError.subscribe(MonoError.java:52)\nat reactor.core.publisher.Mono.subscribe(Mono.java:4218)\nat reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:97)\nat reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onError(FluxMapFuseable.java:134)\nat reactor.core.publisher.MonoCompletionStage.lambda$subscribe$0(MonoCompletionStage.java:80)\nat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)\nat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)\nat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)\nat java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705)\nat java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1692)\nat java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)\nat java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)\nat java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)\nat java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)\nat java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)\nSuppressed: java.lang.Exception: #block terminated with an error\nat reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)\nat reactor.core.publisher.Mono.block(Mono.java:1678)\nat com.azure.security.keyvault.secrets.SecretClient.setSecretWithResponse(SecretClient.java:113)\nat com.azure.security.keyvault.secrets.SecretClient.setSecret(SecretClient.java:93)\nat com.azure.identity.Test.main(Test.java:31)\nCaused by: com.azure.identity.CredentialUnavailableException: DefaultAzureCredential authentication failed. ---> EnvironmentCredential authentication failed. Error Details:AADSTS90002: Tenant '72f988bf-86f1-41af-91ab-2d7cd011db46' not found. This may happen if there are no active subscriptions for the tenant. Check to make sure you have the correct tenant\nID. Check with your subscription administrator.\nTrace ID: 43ffc9ad-f468-4665-ac3a-14a278a98600\nCorrelation ID: e3d3e754-89f0-4839-870c-57e6aba9a5f0\nTimestamp: 2020-06-08 01:49:43Z\nat com.azure.identity.ChainedTokenCredential.lambda$1(ChainedTokenCredential.java:44)\nat reactor.core.publisher.Mono.lambda$onErrorResume$31(Mono.java:3360)\nat reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:88)\n... 12 more\nCaused by: com.microsoft.aad.msal4j.MsalServiceException: AADSTS90002: Tenant '72f988bf-86f1-41af-91ab-2d7cd011db46' not found. This may happen if there are no active subscriptions for the tenant. Check to make sure you have the correct tenant ID. Check with your subscription administrator.\nTrace ID: 43ffc9ad-f468-4665-ac3a-14a278a98600\nCorrelation ID: e3d3e754-89f0-4839-870c-57e6aba9a5f0\nTimestamp: 2020-06-08 01:49:43Z\nat com.microsoft.aad.msal4j.MsalServiceExceptionFactory.fromHttpResponse(MsalServiceExceptionFactory.java:43)\nat com.microsoft.aad.msal4j.TokenRequestExecutor.createAuthenticationResultFromOauthHttpResponse(TokenRequestExecutor.java:81)\nat com.microsoft.aad.msal4j.TokenRequestExecutor.executeTokenRequest(TokenRequestExecutor.java:36)\nat com.microsoft.aad.msal4j.ClientApplicationBase.acquireTokenCommon(ClientApplicationBase.java:92)\nat com.microsoft.aad.msal4j.AcquireTokenByAuthorizationGrantSupplier.execute(AcquireTokenByAuthorizationGrantSupplier.java:52)\nat com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:59)\nat com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:17)\nat java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)\n... 6 more", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r436435674", "createdAt": "2020-06-08T02:53:58Z", "author": {"login": "Luyunmt"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -39,26 +40,24 @@\n         List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n         return Flux.fromIterable(credentials)\n                    .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           return Mono.error(new CredentialUnavailableException(\n+                                   unavailableError + p.getClass().getSimpleName() + \" authentication failed.\",\n+                                   t));\n+                       }\n                        exceptions.add(t);\n                        return Mono.empty();\n                    }), 1)\n                    .next()\n                    .switchIfEmpty(Mono.defer(() -> {\n-\n-                       StringBuilder message = new StringBuilder(\"Tried \"\n-                             + credentials.stream().map(c -> c.getClass().getSimpleName())\n-                                   .collect(Collectors.joining(\", \"))\n-                             + \" but failed to acquire a token for any of them. Please verify the\"\n-                             + \" environment for the credentials\"\n-                             + \" and see more details in the causes below.\");\n-\n                        // Chain Exceptions.\n                        CredentialUnavailableException last = exceptions.get(exceptions.size() - 1);\n                        for (int z = exceptions.size() - 2; z >= 0; z--) {\n                            CredentialUnavailableException current = exceptions.get(z);\n-                           last = new CredentialUnavailableException(current.getMessage(), last);\n+                           last = new CredentialUnavailableException(current.getMessage() + \" \" + last.getMessage(),\n+                                last.getCause());\n                        }\n-                       return Mono.error(new CredentialUnavailableException(message.toString(), last));\n+                       return Mono.error(last);", "state": "SUBMITTED", "replyTo": {"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzNTY4ODYwOQ=="}, "originalCommit": {"oid": "ac8295bf774dec5959a0692f356f60eb36336a1e"}, "originalPosition": 48}]}}, {"id": "MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjczNTY0NTUwOnYy", "diffSide": "RIGHT", "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "isResolved": true, "comments": {"totalCount": 1, "pageInfo": {"startCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0xMlQwMDoxNjo0N1rOGizAnQ==", "endCursor": "Y3Vyc29yOnYyOpK0MjAyMC0wNi0xMlQwMDoxNjo0N1rOGizAnQ==", "hasNextPage": false, "hasPreviousPage": false}, "nodes": [{"id": "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQzOTE0MDUwOQ==", "bodyText": "Why are we wrapping a non Credential Unavailable Exception into a Credential Unavailable Exception ?\nIt should be thrown as it is, to indicate a genuine exception even though credential is available to be used.\nThe distinction will help to debug customer issues too in future.", "url": "https://github.com/Azure/azure-sdk-for-java/pull/9022#discussion_r439140509", "createdAt": "2020-06-12T00:16:47Z", "author": {"login": "g2vinay"}, "path": "sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java", "diffHunk": "@@ -38,27 +39,26 @@\n     public Mono<AccessToken> getToken(TokenRequestContext request) {\n         List<CredentialUnavailableException> exceptions = new ArrayList<>(4);\n         return Flux.fromIterable(credentials)\n-                   .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {\n-                       exceptions.add(t);\n+                   .flatMap(p -> p.getToken(request).onErrorResume(Exception.class, t -> {\n+                       if (!t.getClass().getSimpleName().equals(\"CredentialUnavailableException\")) {\n+                           return Mono.error(new CredentialUnavailableException(", "state": "SUBMITTED", "replyTo": null, "originalCommit": {"oid": "5911c67a9839892c375abed8b4f0d17ed34c1adf"}, "originalPosition": 25}]}}]}}}, "rateLimit": {"limit": 5000, "remaining": 183, "cost": 1, "resetAt": "2021-11-12T11:18:39Z"}}}